### PR TITLE
[DowngradePhp72] Add DowngradePregUnmatchedAsNullConstantRector

### LINKED ISSUE #6013: Undefined constant 'PREG_UNMATCHED_AS_NULL' on rector-prefixed build php 7.1
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | dev-main             |
| Installed as   | latest rector prefixed repository                           |

Current generated rector-prefixed got the following error on php 7.1:

Screen Shot 2021-03-30 at 02 09 04

I can't find define `PREG_UNMATCHED_AS_NULL` on symfony polyfill 7.2. It seems the solution is add custom define constant for it.

@TomasVotruba what do you think?

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #6013

### CODE REVIEW

File: rules/DowngradePhp72/Rector/Function_/DowngradePregUnmatchedAsNullConstantRector.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                    <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public function run()
+    {
+        preg_match('/(a)(b)*(c)/', 'ac', $matches);
+        foreach ($matches as $key => $match) {
+            if ($match === '') {
+                $matches[$key] = null;
+            }
```
Comment (samsonasik) [inline rules/DowngradePhp72/Rector/Function_/DowngradePregUnmatchedAsNullConstantRector.php:59]: To reduce check on variable `key` or `match` already used check, I think we can use self-called anonymous function: 

```php
(function (&$matches) {
    foreach ($matches as $key => $match) {
        if ($matches[$key] === '') {
            $matches[$key] = null;
        }
    }
})($matches);
```

Comment (samsonasik) [reply rules/DowngradePhp72/Rector/Function_/DowngradePregUnmatchedAsNullConstantRector.php:59]: updated to use array_walk_recursive to handle preg_match_all

**Comment by samsonasik** [comment]
It need to check inside if cond as well:

https://github.com/rectorphp/rector-prefixed/blob/5b092a1b8fe62bb78c9a1a626939eacffbe0a5d4/vendor/symfony/string/ByteString.php#L206

**Comment by samsonasik** [comment]
Applied inside if check for usage:

https://github.com/rectorphp/rector-prefixed/blob/5b092a1b8fe62bb78c9a1a626939eacffbe0a5d4/vendor/symfony/string/ByteString.php#L206

**Comment by samsonasik** [comment]
All implemented for `rectorphp/rector-prefixed` usage use cases.

**Comment by samsonasik** [comment]
PHPStan and Unit test error notice seems related with new PHPStan release 3 minutes ago https://github.com/phpstan/phpstan/releases/tag/0.12.83

**Comment by samsonasik** [comment]
I pinned phpstan libs to previous versions:

```diff
-        "phpstan/phpstan": "^0.12.82",
-         "phpstan/phpdoc-parser": "^0.5.2",
-         "phpstan/phpstan-phpunit": "^0.12.18",
+         "phpstan/phpstan": "0.12.82",
+         "phpstan/phpdoc-parser": "0.5.3",
+         "phpstan/phpstan-phpunit": "0.12.18",
```

**Comment by samsonasik** [comment]
The error seems in `rector-symfony` package:

```bash
1) Rector\Symfony\Tests\Rector\ClassMethod\TemplateAnnotationToThisRenderRector\TemplateAnnotationToThisRenderRectorTest::test with data set #5 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
17
tests/Rector/ClassMethod/TemplateAnnotationToThisRenderRector/Fixture/try_catch_array_response.php.inc
18
Failed asserting that string matches format description.
19
--- Expected
20
+++ Actual
21
@@ @@
22
                 'key' => 'value'
23
             ]);
24
         } catch (Throwable $throwable) {
25
-            return new RedirectResponse('here');
26
+            return $this->render('@App/Class_Try_Catch_Array_Response/index.html.twig');
27
         }
28
     }
29
 }
```

I reverted pin phpstan lib versions.

**Comment by samsonasik** [comment]
@TomasVotruba it is ready for review 🎉 The phpunit notice is on `rector-symfony` package.

**Comment by TomasVotruba** [comment]
Thank you :tada: :tada:

**Pull request merged**